bftengine/src/bcstatetransfer/SourceSelector.cpp

```diff
@@ -0,0 +1,111 @@
+// Concord
+//
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include "SourceSelector.hpp"
+
+namespace bftEngine {
+namespace SimpleBlockchainStateTransfer {
+namespace impl {
+
+bool SourceSelector::hasSource() const {
+  return currentReplica_ != NO_REPLICA && sourceSelectionTimeMilli_ > 0;
+}
+
+void SourceSelector::setSourceSelectionTime(uint64_t currTimeMilli) {
+  sourceSelectionTimeMilli_ = currTimeMilli;
+}
+
+void SourceSelector::removeCurrentReplica() {
+  preferredReplicas_.erase(currentReplica_);
+  currentReplica_ = NO_REPLICA;
+}
+
+void SourceSelector::setAllReplicasAsPreferred() { preferredReplicas_ = allOtherReplicas_; }
+
+void SourceSelector::reset() {
+  preferredReplicas_.clear();
+  currentReplica_ = NO_REPLICA;
+  sourceSelectionTimeMilli_ = 0;
+}
+
+bool SourceSelector::isReset() const {
+  return preferredReplicas_.empty() && currentReplica_ == NO_REPLICA && sourceSelectionTimeMilli_ == 0;
+}
+
+bool SourceSelector::retransmissionTimeoutExpired(uint64_t currTimeMilli) const {
+  // TODO(GG): TBD - compute dynamically
+  return timeSinceSourceSelectedMilli(currTimeMilli) > retransmissionTimeoutMilli_;
+}
+
+uint64_t SourceSelector::timeSinceSourceSelectedMilli(uint64_t currTimeMilli) const {
+  return ((currentReplica_ == NO_REPLICA) || (currTimeMilli < sourceSelectionTimeMilli_))
+             ? 0
+             : (currTimeMilli - sourceSelectionTimeMilli_);
+}
+
+// Replace the source if necessary.
+//
+// Return true if the source was updated, false otherwise.
+bool SourceSelector::updateSource(bool badDataFromCurrentSource, uint64_t currTimeMilli) {
+  if (shouldReplaceSource(currTimeMilli, badDataFromCurrentSource, sourceReplacementTimeoutMilli_)) {
+    if (currentReplica_ != NO_REPLICA) {
+      preferredReplicas_.erase(currentReplica_);
+    }
+    if (preferredReplicas_.size() == 0) {
```

Comment:
-> if (preferredReplicas_.empty())

Reply:
> The changes look reasonable. Could we have a test verifying solution?

The behavior is already tested via the state transfer and state transfer persistence system tests. This was just a straight extraction to a class. The algorithm didn't change.